src/components/ChangesNotAppliedModal.jsx

```diff
@@ -1,37 +1,51 @@
 import React from 'react';
+import { useSelector, useDispatch } from 'react-redux';
 import PropTypes from 'prop-types';
 import {
   Modal, Typography, Space, Button,
 } from 'antd';
+import { useRouter } from 'next/router';
+import { discardChangedQCFilters, navigateFromProcessingTo } from '../redux/actions/experimentSettings';
+import { runPipeline } from '../redux/actions/pipeline';
 
 import { getUserFriendlyQCStepName } from '../utils/qcSteps';
 
 const { Text } = Typography;
 
 const ChangesNotAppliedModal = (props) => {
-  const {
-    steps, visible, onRun, onDiscard, onCancel,
-  } = props;
+  const { experimentId } = props;
+
+  const router = useRouter();
+  const dispatch = useDispatch();
+
+  const { changedQCFilters, navigationPath } = useSelector((state) => state.experimentSettings.processing.meta);
 
   return (
     <Modal
-      visible={visible}
+      visible={navigationPath}
       title='Changes not applied'
-      onCancel={onCancel}
+      onCancel={() => dispatch(navigateFromProcessingTo(''))}
       footer={(
         <Space size='large' style={{ display: 'flex', justifyContent: 'center' }}>
           <Button
             type='primary'
             key='run'
-            onClick={onRun}
+            onClick={() => {
```

Comment:
Why was this moved into the modal? FilesUploadModal doesn't trigger the upload for example

Reply:
This modal directly listens to the `navigationPath` to get rendered or not. This is so because this modal needs to be called from both the navigation links and also the finish qc button. Because it listens directly to the store, I thought it would be ok for its actions to also be directly related to the store instead of taking functions from its parent component.

Reply:
The thing is that no other modal is doing this, all of them work through callbacks

Reply:
Perhaps could we wrap this ChangesNotAppliedModal in a `NavigationHandler` that handles the navigation requests

Reply:
and move the callbacks back outside, with this navigation component handling those callbacks via dispatches.

Reply:
Hmm... yea. Ok, I'll move the actions to the parent components,

Reply:
Does that mean the other navigation will also route throught this middleware component?

Reply:
something like our notifications manager